### PR TITLE
Rule S6302: Policies granting all privileges are security-sensitive for PolicyStatements

### PR DESCRIPTION
**Pull request by nils-werner-sonarsource**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/cdk/PrivilegePolicyCheck.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    if (action.hasExpression(isStringLiteral())) {
+      action.addIssueIf(isString("*"), MESSAGE, secondaryLocation(effect));
+    } else {
+      CdkUtils.getList(action).ifPresent(actions -> checkWildcardAction(ctx, actions, effect));
+    }
+  }
+
+  private static void checkWildcardAction(SubscriptionContext ctx, ListLiteral list, ExpressionFlow effect) {
+    CdkUtils.getListElements(ctx, list)
+      .stream()
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/cdk/PrivilegePolicyCheck.java:96]:
```suggestion
    CdkUtils.getListElements(ctx, list).stream()
```

File: python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java [hunk 2]
```diff
@@ -206,5 +255,9 @@ private ResolvedKeyValuePair(ExpressionFlow key, ExpressionFlow value) {
     static ResolvedKeyValuePair build(SubscriptionContext ctx, KeyValuePair pair) {
       return new ResolvedKeyValuePair(ExpressionFlow.build(ctx, pair.key()), ExpressionFlow.build(ctx, pair.value()));
     }
+
+    public ExpressionFlow value() {
```
Comment (mstachniuk) [inline python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java:259]: I understand the reason behind this method (to use `.map(ResolvedKeyValuePair::value)`) but `key` and `value` are package private and `.value` is used in many places. I think there is a god practice to use always either field access or getter/method access (for consistency in the code), but never both at the same time. WDYT?

File: python-checks/src/test/java/org/sonar/python/checks/cdk/PrivilegePolicyCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.sonar.plugins.python.api.PythonCheck;
+import org.sonar.python.checks.utils.PythonCheckVerifier;
+
+public class PrivilegePolicyCheckTest {
+
+  PythonCheck check = new PrivilegePolicyCheck();
+
+  @Test
+  public void policyStatement() {
+    PythonCheckVerifier.verify("src/test/resources/checks/cdk/PrivilegePolicyCheck/policyStatement.py", check);
```
Comment (mstachniuk) [inline python-checks/src/test/java/org/sonar/python/checks/cdk/PrivilegePolicyCheckTest.java:32]: PrivilegePolicyCheck - it should be snake case

**Review by mstachniuk — APPROVED** [review]
LGTM. I put some comments, most of them are newbie questions

File: python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java [hunk 4]
```diff
@@ -101,6 +105,50 @@ protected static Optional<ExpressionFlow> getArgument(SubscriptionContext ctx, C
...
+      .flatMap(Optional::stream)
+      .collect(Collectors.toList());
+  }
+
+  public static List<ResolvedKeyValuePair> resolveDictionary(SubscriptionContext ctx, DictionaryLiteral dict) {
+    return dict.elements().stream()
+      .map(e -> CdkUtils.getKeyValuePair(ctx, e))
+      .flatMap(Optional::stream)
+      .collect(Collectors.toUnmodifiableList());
+  }
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java:150]: I had the same thought, but was honestly too lazy. I will make up for it.

File: python-checks/src/test/resources/checks/cdk/PrivilegePolicyCheck/fromJson.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+from aws_cdk import aws_iam as iam
+
+iam.PolicyStatement.from_json({
+    "Effect": "Allow",
+ #  ^^^^^^^^^^^^^^^^^> {{Related effect}}
+    "Action": ["*"],  # Noncompliant {{Make sure granting all privileges is safe here.}}
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply python-checks/src/test/resources/checks/cdk/PrivilegePolicyCheck/fromJson.py:6]: Should already work. I will add a testcase.

File: python-checks/src/main/java/org/sonar/python/checks/cdk/PrivilegePolicyCheck.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+public class PrivilegePolicyCheck extends AbstractCdkResourceCheck {
+
+  private static final String MESSAGE = "Make sure granting all privileges is safe here.";
+  private static final String SECONDARY_MESSAGE = "Related effect";
+
+  @Override
+  protected void registerFqnConsumer() {
+    checkFqn("aws_cdk.aws_iam.PolicyStatement", (ctx, call) -> {
+      @Nullable ExpressionFlow effect = CdkUtils.getArgument(ctx, call, "effect").orElse(null);
+      CdkUtils.getArgumentAsList(ctx, call, "actions").ifPresent(actions -> {
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/cdk/PrivilegePolicyCheck.java:50]: Unfortunately, we do not have a convention for this. We should address times.

File: python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java (hunk 2 above)
Comment (nils-werner-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/cdk/CdkUtils.java:259]: I agree. Let's keep the field access for now

**Pull request merged**